File: app/src/main/java/com/example/groxweatherapp/api/WeatherApiClient.java
package com.example.groxweatherapp.api;

import android.support.annotation.IntDef;
import com.example.groxweatherapp.model.Weather;
import com.example.groxweatherapp.model.WeatherResponse;
import java.lang.annotation.Retention;
import java.util.ArrayList;
import java.util.List;
import java.util.Random;
import java.util.concurrent.ThreadLocalRandom;
import java.util.concurrent.TimeUnit;
import rx.Observable;

import static java.lang.annotation.RetentionPolicy.SOURCE;

public class WeatherApiClient {
    private static final int SEED = 7;
    private static final int ERROR_RATE = 5;
    public static final int LATENCY_IN_MS = 1000;
    private static Random random = new Random(SEED);
    private static final String ERROR_MESSAGE = "Client Exception";

    private static final String DAYS[] = {
        "TODAY", "SUN", "MON", "TUES", "THURS", "FRI", "SAT"
    };
    private static final String DESCRIPTION[] = {
        "SUNNY", "OVERCAST", "CLOUDY", "PARTLY SUNNY", "PARTLY CLOUD"
    };
    private static final String DATES[] = {
        "MAR 1", "MAR 2", "MAR 3", "MAR 4", "MAR 5", "MAR 6", "MAR 7", "MAR 8", "MAR 9", "MAR 10"
    };

    public static int MAX = 100;
    public static int MIN = 40;

    @Retention(SOURCE)
    @IntDef({TODAY, FIVE_DAY, TEN_DAY})
    public @interface RequestMode {}

    public static final int TODAY = 0;
    public static final int FIVE_DAY = 1;
    public static final int TEN_DAY = 2;

    public Observable<WeatherResponse> getWeather(@RequestMode int requestMode) {
        WeatherResponse weatherResponse = new WeatherResponse();
        weatherResponse.weathers = createWeatherForRequest(requestMode);
        Observable<WeatherResponse> observable = Observable.just(weatherResponse);
        if (random.nextInt() % ERROR_RATE == 0) {
            observable = Observable.error(new RuntimeException(ERROR_MESSAGE));
        }

        return observable.delay(LATENCY_IN_MS, TimeUnit.MILLISECONDS);
    }

    private List<Weather> createWeatherForRequest(@RequestMode int requestMode) {
        final List<Weather> weatherList = new ArrayList<>();

        switch (requestMode) {
            case TODAY:
                return createAndAddWeatherToList(1, weatherList);
            case FIVE_DAY:
                return createAndAddWeatherToList(5, weatherList);
            case TEN_DAY:
                return createAndAddWeatherToList(10, weatherList);
        }

        return weatherList;
    }

    private List<Weather> createAndAddWeatherToList(int numOfDays, List<Weather> weatherList) {
        for (int i = 0; i < numOfDays; i++) {
            weatherList.add(createWeather(DAYS[i % (DAYS.length - 1)] + " " + DATES[i]));
        }
        return weatherList;
    }

    private Weather createWeather(String date) {
        final Weather weather = new Weather();
        weather.date = date;
        int index = getRandom(MIN, MAX) % (DESCRIPTION.length - 1);
        weather.description = DESCRIPTION[index];
        final int maxTemp = getRandom(MIN, MAX);
        weather.temperatureMax = String.valueOf(maxTemp) + (char) 0x00B0;
        weather.temperatureMin = String.valueOf(getRandom(MIN, maxTemp + 1)) + (char) 0x00B0;
        return weather;
    }

    private int getRandom(int min, int max) {
        return ThreadLocalRandom.current().nextInt(min, max);
    }
}

File: app/src/main/java/com/example/groxweatherapp/api/WeatherApiClient.kt
package com.example.groxweatherapp.api

import android.support.annotation.IntDef
import com.example.groxweatherapp.model.Weather
import com.example.groxweatherapp.model.WeatherResponse
import rx.Observable
import java.lang.annotation.Retention
import java.lang.annotation.RetentionPolicy.SOURCE
import java.util.*
import java.util.concurrent.ThreadLocalRandom
import java.util.concurrent.TimeUnit

class WeatherApiClient {

    fun getWeather(@ForecastMode forecastMode: Int): Observable<WeatherResponse> {
        val weathers: List<Weather>  = createWeatherForRequest(forecastMode)
        val weatherResponse = WeatherResponse(weathers)
        var observable = Observable.just(weatherResponse)
        if (random.nextInt() % ERROR_RATE == 0) {
            observable = Observable.error(RuntimeException(ERROR_MESSAGE))
        }

        return observable.delay(LATENCY_IN_MS.toLong(), TimeUnit.MILLISECONDS)
    }

    private fun createWeatherForRequest(@ForecastMode forecastMode: Int): List<Weather> {
        val weatherList = ArrayList<Weather>()

        when (forecastMode) {
            TODAY -> return createAndAddWeatherToList(1, weatherList)
            FIVE_DAY -> return createAndAddWeatherToList(5, weatherList)
            TEN_DAY -> return createAndAddWeatherToList(10, weatherList)
        }

        return weatherList
    }

    private fun createAndAddWeatherToList(numOfDays: Int, weatherList: MutableList<Weather>): List<Weather> {
        for (i in 0 until numOfDays) {
            weatherList.add(createWeather(DAYS[i % (DAYS.size - 1)] + " " + DATES[i]))
        }
        return weatherList
    }

    private fun createWeather(date: String): Weather {
        val maxTemp = getRandom(MIN, MAX)
        val temperatureMax = maxTemp.toString() + 0x00B0.toChar()
        val temperatureMin = getRandom(MIN, maxTemp + 1).toString() + 0x00B0.toChar()
        val description = DESCRIPTION[getRandom(MIN, MAX) % (DESCRIPTION.size - 1)]
        return Weather(date, temperatureMax, temperatureMin, description)
    }

    private fun getRandom(min: Int, max: Int): Int {
        return ThreadLocalRandom.current().nextInt(min, max)
    }

    companion object {
        private const val SEED = 7
        private const val ERROR_RATE = 5
        private const val LATENCY_IN_MS = 1000
        private const val ERROR_MESSAGE = "Client Exception"

        private const val MAX = 100
        private const val MIN = 40
        private val random = Random(SEED.toLong())

        private val DAYS = arrayOf("TODAY", "SUN", "MON", "TUES", "THURS", "FRI", "SAT")
        private val DESCRIPTION = arrayOf("SUNNY", "OVERCAST", "CLOUDY", "PARTLY SUNNY", "PARTLY CLOUD")
        private val DATES = arrayOf("MAR 1", "MAR 2", "MAR 3", "MAR 4", "MAR 5", "MAR 6", "MAR 7", "MAR 8", "MAR 9", "MAR 10")

        @IntDef(TODAY.toLong(), FIVE_DAY.toLong(), TEN_DAY.toLong())
        @Retention(SOURCE)
        annotation class ForecastMode

        const val TODAY = 0
        const val FIVE_DAY = 1
        const val TEN_DAY = 2
    }
}
